compiler/src/bytecode.rs

```diff
@@ -21,120 +32,203 @@ impl Bytecode {
         self.len() == 0
     }
 
+    /// emits a signed 64 bits integer
+    pub fn emit_int(&mut self, value: i64) {
+        self.bytes.extend(value.to_be_bytes());
+    }
+
+    /// emits an unsigned byte
+    pub fn emit_byte(&mut self, value: u8) {
+        self.bytes.push(value);
+    }
+
+    /// emits an unsigned 32 bits integer
+    pub fn emit_u32(&mut self, value: u32) {
+        self.bytes.extend(value.to_be_bytes());
+    }
+
+    /// emits a signed 64 bits float
+    pub fn emit_float(&mut self, value: f64) {
+        self.bytes.extend(value.to_be_bytes());
+    }
+
+    /// emits a constant pool reference, which is an unsigned 32 bits integer
+    pub fn emit_constant_ref(&mut self, constant: u32) {
+        self.emit_u32(constant);
+    }
+
+    /// Fills an instruction pointer at given instruction pointer in the byte array
+    pub fn patch_u32_placeholder(&mut self, placeholder: Placeholder, value: u32) {
+        let pos = placeholder.pos as usize;
+        self.bytes[pos..pos + size_of::<u32>()].copy_from_slice(&value.to_be_bytes())
+    }
+
+    pub fn emit_instruction_pointer(&mut self, ip: usize) {
+        self.bytes.extend(ip.to_be_bytes());
+    }
+
+    /// expands the byte vector to let a placeholder of the given size,
+    /// returning the position of the placeholder in the vector
+    pub fn emit_u32_placeholder(&mut self) -> Placeholder {
+        let pos = self.bytes.len();
+        self.bytes.resize(pos + size_of::<u32>(), 0);
+        Placeholder { pos: pos as u32 }
+    }
+}
+
+/// This structure is a [Bytecode] wrapper and is used
+/// to emit bytecode instructions.
+pub struct Instructions<'a> {
+    pub bytecode: &'a mut Bytecode,
+    // offset of where this instruction set starts in the given bytecode
+    pub ip_offset: usize,
+}
+
+impl<'a> Instructions<'a> {
+    pub fn wrap(bytecode: &'a mut Bytecode) -> Self {
+        Self {
+            ip_offset: bytecode.len(),
+            bytecode,
+        }
+    }
+
     pub fn emit_code(&mut self, code: Opcode) {
-        self.bytes.push(code as u8);
+        self.bytecode.emit_byte(code as u8)
+    }
+
+    pub fn emit_pop(&mut self, size: ValueStackSize) {
+        let pop_opcode = match size {
+            ValueStackSize::Zero => panic!("Cannot pop zero sized types"),
```

Comment:
```suggestion
            ValueStackSize::Zero => return,
```
Writing `fun a() = {}; a()` should be valid.